app/apollo/resolvers/cluster.js

```diff
@@ -434,14 +501,21 @@ const clusterResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating`);
+        const cluster = await models.Cluster.findOne({
+          org_id,
+          cluster_id
+        }).lean({ virtuals: true });
 
-        await validAuth(me, org_id, ACTIONS.DETACH, TYPES.CLUSTER, queryName, context);
+        if(!cluster){
+          throw new NotFoundError(context.req.t('Could not find the cluster with Id {{clusterId}}.', {'clusterId':cluster_id}), context);
+        }
 
         validateString( 'org_id', org_id );
         validateString( 'cluster_id', cluster_id );
 
-        logger.info({req_id, user, org_id, cluster_id}, `${queryName} saving`);
+        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating`);
+        await validAuth(me, org_id, ACTIONS.DETACH, TYPES.CLUSTER, queryName, context, [cluster_id, cluster.registration.name || cluster.name]);
+        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating - authorized`);
```

Comment:
Querying the DB should come after basic input validation.

NotFoundError should be thrown _after_ auth check so as to not 'leak' info.  Otherwise a user without access could use whether they get a NotFoundError or a NotAuthorized error to determine if a cluster exists, even without access.

Probably something like:
```
        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating`);

        validateString( 'org_id', org_id );
        validateString( 'cluster_id', cluster_id );

        const cluster = await models.Cluster.findOne({
          org_id,
          cluster_id
        }).lean({ virtuals: true });
        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating - found: ${!!cluster}`); // the `!!` ensures it logs true or false rather than the actual cluster object.

        await validAuth(me, org_id, ACTIONS.DETACH, TYPES.CLUSTER, queryName, context, cluster ? [cluster_id, cluster.registration.name || cluster.name] : [cluster_id] );  // Check auth with cluster id and name if the cluster exists in the DB, else just use the cluster_id
        logger.info({req_id, user, org_id, cluster_id}, `${queryName} validating - authorized`);

        // If user is authorized but cluster does not exist, throw NotFoundError
        if(!cluster){
          throw new NotFoundError(context.req.t('Could not find the cluster with Id {{clusterId}}.', {'clusterId':cluster_id}), context);
        }
```

Reply:
Makes sense, pushed up